Wireplumber module does not build with Wireplumber 0.5 API

Fedora rawhide has updated wireplumber to 0.4.81 (0.5 API release preview) which seems to have altered use of the GLib logging API and some initialisation callbacks.

Bulding with Clang:

```
[65/93] Compiling C++ object waybar.p/src_modules_wireplumber.cpp.o
FAILED: waybar.p/src_modules_wireplumber.cpp.o
/usr/bin/clang++ -Iwaybar.p -I. -I.. -I../include -I../subprojects/spdlog-1.11.0/include -I../subprojects/cava-0.9.1/include -Iresources/icons -Iprotocol -I/usr/local/include -I/usr/include/sigc++-2.0 -I/usr/lib64/sigc++-2.0/include -I/usr/include/gtkmm-3.0 -I/usr/lib64/gtkmm-3.0/include -I/usr/include/atkmm-1.6 -I/usr/lib64/atkmm-1.6/include -I/usr/include/gtk-3.0/unix-print -I/usr/include/gdkmm-3.0 -I/usr/lib64/gdkmm-3.0/include -I/usr/include/gtk-3.0 -I/usr/include/pango-1.0 -I/usr/include/giomm-2.4 -I/usr/lib64/giomm-2.4/include -I/usr/include/pangomm-1.4 -I/usr/lib64/pangomm-1.4/include -I/usr/include/glibmm-2.4 -I/usr/lib64/glibmm-2.4/include -I/usr/include/cairo -I/usr/include/cairomm-1.0 -I/usr/lib64/cairomm-1.0/include -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/dbus-1.0 -I/usr/lib64/dbus-1.0/include -I/usr/include/atk-1.0 -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/at-spi2-atk/2.0 -I/usr/include/cloudproviders -I/usr/include/webp -I/usr/include/at-spi-2.0 -I/usr/include/blkid -I/usr/include/gio-unix-2.0 -I/usr/include/libmount -I/usr/include/pixman-1 -I/usr/include/libxml2 -I/usr/include/fribidi -I/usr/include/sysprof-6 -I/usr/include/libpng16 -I/usr/include/libdbusmenu-gtk3-0.4 -I/usr/include/libdbusmenu-glib-0.4 -I/usr/include/libnl3 -I/usr/include/libupower-glib -I/usr/include/pipewire-0.3 -I/usr/include/spa-0.2 -I/usr/include/playerctl -I/usr/include/wireplumber-0.5 -I/usr/include/libevdev-1.0 -I/usr/local/include/gtk-layer-shell -I/usr/include/SDL2 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++20 -O3 '-DVERSION="0.9.24-139-g1dce607c (branch '"'"'master'"'"')"' -DHAVE_LANGINFO_1STDAY '-DSYSCONFDIR="//usr/local/etc"' -DHAVE_CPU_LINUX -DHAVE_MEMORY_LINUX -DHAVE_SYSTEMD_MONITOR -DHAVE_SWAY -DHAVE_WLR -DHAVE_RIVER -DHAVE_DWL -DHAVE_HYPRLAND -DHAVE_LIBNL -DHAVE_GAMEMODE -DHAVE_UPOWER -DHAVE_PIPEWIRE -DHAVE_MPRIS -DHAVE_LIBPULSE -DHAVE_LIBJACK -DHAVE_LIBWIREPLUMBER -DHAVE_DBUSMENU -DHAVE_LIBUDEV -DHAVE_LIBEVDEV -DHAVE_LIBINPUT -DHAVE_GTK_LAYER_SHELL -DHAVE_GIO_UNIX -DHAVE_CHRONO_TIMEZONES -DHAVE_LIBCAVA -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -pthread -D_REENTRANT -DWITH_GZFILEOP -DSPDLOG_FMT_EXTERNAL -MD -MQ waybar.p/src_modules_wireplumber.cpp.o -MF waybar.p/src_modules_wireplumber.cpp.o.d -o waybar.p/src_modules_wireplumber.cpp.o -c ../src/modules/wireplumber.cpp
In file included from ../src/modules/wireplumber.cpp:1:
In file included from ../include/modules/wireplumber.hpp:9:
In file included from ../include/ALabel.hpp:4:
In file included from /usr/include/gtkmm-3.0/gtkmm/label.h:27:
In file included from /usr/include/gtkmm-3.0/gtkmm/misc.h:27:
In file included from /usr/include/gtkmm-3.0/gtkmm/widget.h:31:
In file included from /usr/include/pangomm-1.4/pangomm/context.h:29:
In file included from /usr/include/glibmm-2.4/glibmm/object.h:30:
/usr/include/glibmm-2.4/glibmm/wrap.h:92:7: error: use of undeclared identifier 'WP_LOCAL_LOG_TOPIC'
   92 |       g_warning("Glib::wrap_auto_interface(): The C++ instance (%s) does not dynamic_cast to the "
      |       ^
/usr/include/glib-2.0/glib/gmessages.h:362:32: note: expanded from macro 'g_warning'
  362 | #define g_warning(...)  g_log (G_LOG_DOMAIN,         \
      |                                ^
/usr/include/wireplumber-0.5/wp/log.h:63:23: note: expanded from macro 'G_LOG_DOMAIN'
   63 | #define G_LOG_DOMAIN (WP_LOCAL_LOG_TOPIC->topic_name)
      |                       ^
In file included from ../src/modules/wireplumber.cpp:1:
In file included from ../include/modules/wireplumber.hpp:9:
In file included from ../include/ALabel.hpp:4:
In file included from /usr/include/gtkmm-3.0/gtkmm/label.h:27:
In file included from /usr/include/gtkmm-3.0/gtkmm/misc.h:27:
In file included from /usr/include/gtkmm-3.0/gtkmm/widget.h:54:
/usr/include/giomm-2.4/giomm/actiongroup.h:702:3: error: use of undeclared identifier 'WP_LOCAL_LOG_TOPIC'
  702 |   g_return_if_fail(
      |   ^
/usr/include/glib-2.0/glib/gmessages.h:652:35: note: expanded from macro 'g_return_if_fail'
  652 |         g_return_if_fail_warning (G_LOG_DOMAIN, \
      |                                   ^
/usr/include/wireplumber-0.5/wp/log.h:63:23: note: expanded from macro 'G_LOG_DOMAIN'
   63 | #define G_LOG_DOMAIN (WP_LOCAL_LOG_TOPIC->topic_name)
      |                       ^
In file included from ../src/modules/wireplumber.cpp:1:
In file included from ../include/modules/wireplumber.hpp:9:
In file included from ../include/ALabel.hpp:4:
In file included from /usr/include/gtkmm-3.0/gtkmm/label.h:27:
In file included from /usr/include/gtkmm-3.0/gtkmm/misc.h:27:
In file included from /usr/include/gtkmm-3.0/gtkmm/widget.h:54:
/usr/include/giomm-2.4/giomm/actiongroup.h:722:3: error: use of undeclared identifier 'WP_LOCAL_LOG_TOPIC'
  722 |   g_return_if_fail(
      |   ^
/usr/include/glib-2.0/glib/gmessages.h:652:35: note: expanded from macro 'g_return_if_fail'
  652 |         g_return_if_fail_warning (G_LOG_DOMAIN, \
      |                                   ^
/usr/include/wireplumber-0.5/wp/log.h:63:23: note: expanded from macro 'G_LOG_DOMAIN'
   63 | #define G_LOG_DOMAIN (WP_LOCAL_LOG_TOPIC->topic_name)
      |                       ^
../src/modules/wireplumber.cpp:258:8: error: no matching function for call to 'wp_core_load_component'
  258 |   if (!wp_core_load_component(wp_core_, "libwireplumber-module-default-nodes-api", "module", NULL,
      |        ^~~~~~~~~~~~~~~~~~~~~~
/usr/include/wireplumber-0.5/wp/component-loader.h:44:6: note: candidate function not viable: requires 8 arguments, but 5 were provided
   44 | void wp_core_load_component (WpCore * self, const gchar * component,
      |      ^                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   45 |     const gchar * type, WpSpaJson * args, const gchar * provides,
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   46 |     GCancellable * cancellable, GAsyncReadyCallback callback, gpointer data);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/modules/wireplumber.cpp:258:7: error: invalid argument type 'void' to unary expression
  258 |   if (!wp_core_load_component(wp_core_, "libwireplumber-module-default-nodes-api", "module", NULL,
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  259 |                               &error)) {
      |                               ~~~~~~~
../src/modules/wireplumber.cpp:263:8: error: no matching function for call to 'wp_core_load_component'
  263 |   if (!wp_core_load_component(wp_core_, "libwireplumber-module-mixer-api", "module", NULL,
      |        ^~~~~~~~~~~~~~~~~~~~~~
/usr/include/wireplumber-0.5/wp/component-loader.h:44:6: note: candidate function not viable: requires 8 arguments, but 5 were provided
   44 | void wp_core_load_component (WpCore * self, const gchar * component,
      |      ^                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   45 |     const gchar * type, WpSpaJson * args, const gchar * provides,
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   46 |     GCancellable * cancellable, GAsyncReadyCallback callback, gpointer data);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/modules/wireplumber.cpp:263:7: error: invalid argument type 'void' to unary expression
  263 |   if (!wp_core_load_component(wp_core_, "libwireplumber-module-mixer-api", "module", NULL,
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  264 |                               &error)) {
      |                               ~~~~~~~
7 errors generated.
```
The `wp_core_load_component` can be hacked around by ignoring errors and passing `NULL` for the ready callbacks, but these could probably be handled robustly, and the logging errors seem to be all in included GLib headers so probably need a correct `WP_LOCAL_LOG_TOPIC` defined.

--- a/include/modules/wireplumber.hpp
+++ b/include/modules/wireplumber.hpp
@@ -17,12 +17,15 @@ class Wireplumber : public ALabel {
   auto update() -> void override;
 
  private:
-  void loadRequiredApiModules();
+  void asyncLoadRequiredApiModules();
   void prepare();
   void activatePlugins();
   static void updateVolume(waybar::modules::Wireplumber* self, uint32_t id);
   static void updateNodeName(waybar::modules::Wireplumber* self, uint32_t id);
   static void onPluginActivated(WpObject* p, GAsyncResult* res, waybar::modules::Wireplumber* self);
+  static void onDefaultNodesApiLoaded(WpObject* p, GAsyncResult* res,
+                                      waybar::modules::Wireplumber* self);
+  static void onMixerApiLoaded(WpObject* p, GAsyncResult* res, waybar::modules::Wireplumber* self);
   static void onObjectManagerInstalled(waybar::modules::Wireplumber* self);
   static void onMixerChanged(waybar::modules::Wireplumber* self, uint32_t id);
   static void onDefaultNodesApiChanged(waybar::modules::Wireplumber* self);

--- a/meson.build
+++ b/meson.build
@@ -92,7 +92,7 @@ libevdev = dependency('libevdev', required: get_option('libevdev'))
 libmpdclient = dependency('libmpdclient', required: get_option('mpd'))
 xkbregistry = dependency('xkbregistry')
 libjack = dependency('jack', required: get_option('jack'))
-libwireplumber = dependency('wireplumber-0.4', required: get_option('wireplumber'))
+libwireplumber = dependency('wireplumber-0.5', required: get_option('wireplumber'))
 
 libsndio = compiler.find_library('sndio', required: get_option('sndio'))
 if libsndio.found()

--- a/src/modules/wireplumber.cpp
+++ b/src/modules/wireplumber.cpp
@@ -18,31 +18,24 @@ waybar::modules::Wireplumber::Wireplumber(const std::string& id, const Json::Val
       min_step_(0.0),
       node_id_(0) {
   wp_init(WP_INIT_PIPEWIRE);
-  wp_core_ = wp_core_new(NULL, NULL);
+  wp_core_ = wp_core_new(nullptr, nullptr, nullptr);
   apis_ = g_ptr_array_new_with_free_func(g_object_unref);
   om_ = wp_object_manager_new();
 
   prepare();
 
-  loadRequiredApiModules();
+  spdlog::debug("[{}]: connecting to pipewire...", name_);
 
-  spdlog::debug("[{}]: connecting to pipewire...", this->name_);
-
-  if (!wp_core_connect(wp_core_)) {
-    spdlog::error("[{}]: Could not connect to PipeWire", this->name_);
+  if (wp_core_connect(wp_core_) == 0) {
+    spdlog::error("[{}]: Could not connect to PipeWire", name_);
     throw std::runtime_error("Could not connect to PipeWire\n");
   }
 
-  spdlog::debug("[{}]: connected!", this->name_);
+  spdlog::debug("[{}]: connected!", name_);
 
   g_signal_connect_swapped(om_, "installed", (GCallback)onObjectManagerInstalled, this);
 
-  activatePlugins();
-
-  dp.emit();
-
-  event_box_.add_events(Gdk::SCROLL_MASK | Gdk::SMOOTH_SCROLL_MASK);
-  event_box_.signal_scroll_event().connect(sigc::mem_fun(*this, &Wireplumber::handleScroll));
+  asyncLoadRequiredApiModules();
 }
 
 waybar::modules::Wireplumber::~Wireplumber() {
@@ -63,32 +56,36 @@ void waybar::modules::Wireplumber::updateNodeName(waybar::modules::Wireplumber* 
     return;
   }
 
-  auto proxy = static_cast<WpProxy*>(wp_object_manager_lookup(
-      self->om_, WP_TYPE_GLOBAL_PROXY, WP_CONSTRAINT_TYPE_G_PROPERTY, "bound-id", "=u", id, NULL));
+  auto* proxy = static_cast<WpProxy*>(wp_object_manager_lookup(self->om_, WP_TYPE_GLOBAL_PROXY,
+                                                               WP_CONSTRAINT_TYPE_G_PROPERTY,
+                                                               "bound-id", "=u", id, nullptr));
 
-  if (!proxy) {
+  if (proxy == nullptr) {
     auto err = fmt::format("Object '{}' not found\n", id);
     spdlog::error("[{}]: {}", self->name_, err);
     throw std::runtime_error(err);
   }
 
   g_autoptr(WpProperties) properties =
-      WP_IS_PIPEWIRE_OBJECT(proxy) ? wp_pipewire_object_get_properties(WP_PIPEWIRE_OBJECT(proxy))
-                                   : wp_properties_new_empty();
-  g_autoptr(WpProperties) global_p = wp_global_proxy_get_global_properties(WP_GLOBAL_PROXY(proxy));
+      WP_IS_PIPEWIRE_OBJECT(proxy) != 0
+          ? wp_pipewire_object_get_properties(WP_PIPEWIRE_OBJECT(proxy))
+          : wp_properties_new_empty();
+  g_autoptr(WpProperties) globalP = wp_global_proxy_get_global_properties(WP_GLOBAL_PROXY(proxy));
   properties = wp_properties_ensure_unique_owner(properties);
-  wp_properties_add(properties, global_p);
-  wp_properties_set(properties, "object.id", NULL);
-  auto nick = wp_properties_get(properties, "node.nick");
-  auto description = wp_properties_get(properties, "node.description");
+  wp_properties_add(properties, globalP);
+  wp_properties_set(properties, "object.id", nullptr);
+  const auto* nick = wp_properties_get(properties, "node.nick");
+  const auto* description = wp_properties_get(properties, "node.description");
 
-  self->node_name_ = nick ? nick : description ? description : "Unknown node name";
+  self->node_name_ = nick != nullptr          ? nick
+                     : description != nullptr ? description
+                                              : "Unknown node name";
   spdlog::debug("[{}]: Updating node name to: {}", self->name_, self->node_name_);
 }
 
 void waybar::modules::Wireplumber::updateVolume(waybar::modules::Wireplumber* self, uint32_t id) {
   spdlog::debug("[{}]: updating volume", self->name_);
-  GVariant* variant = NULL;
+  GVariant* variant = nullptr;
 
   if (!isValidNodeId(id)) {
     spdlog::error("[{}]: '{}' is not a valid node ID. Ignoring volume update.", self->name_, id);
@@ -97,7 +94,7 @@ void waybar::modules::Wireplumber::updateVolume(waybar::modules::Wireplumber* se
 
   g_signal_emit_by_name(self->mixer_api_, "get-volume", id, &variant);
 
-  if (!variant) {
+  if (variant == nullptr) {
     auto err = fmt::format("Node {} does not support volume\n", id);
     spdlog::error("[{}]: {}", self->name_, err);
     throw std::runtime_error(err);
@@ -115,9 +112,9 @@ void waybar::modules::Wireplumber::onMixerChanged(waybar::modules::Wireplumber* 
   spdlog::debug("[{}]: (onMixerChanged) - id: {}", self->name_, id);
 
   g_autoptr(WpNode) node = static_cast<WpNode*>(wp_object_manager_lookup(
-      self->om_, WP_TYPE_NODE, WP_CONSTRAINT_TYPE_G_PROPERTY, "bound-id", "=u", id, NULL));
+      self->om_, WP_TYPE_NODE, WP_CONSTRAINT_TYPE_G_PROPERTY, "bound-id", "=u", id, nullptr));
 
-  if (!node) {
+  if (node == nullptr) {
     spdlog::warn("[{}]: (onMixerChanged) - Object with id {} not found", self->name_, id);
     return;
   }
@@ -140,49 +137,49 @@ void waybar::modules::Wireplumber::onMixerChanged(waybar::modules::Wireplumber* 
 void waybar::modules::Wireplumber::onDefaultNodesApiChanged(waybar::modules::Wireplumber* self) {
   spdlog::debug("[{}]: (onDefaultNodesApiChanged)", self->name_);
 
-  uint32_t default_node_id;
-  g_signal_emit_by_name(self->def_nodes_api_, "get-default-node", "Audio/Sink", &default_node_id);
+  uint32_t defaultNodeId;
+  g_signal_emit_by_name(self->def_nodes_api_, "get-default-node", "Audio/Sink", &defaultNodeId);
 
-  if (!isValidNodeId(default_node_id)) {
+  if (!isValidNodeId(defaultNodeId)) {
     spdlog::warn("[{}]: '{}' is not a valid node ID. Ignoring node change.", self->name_,
-                 default_node_id);
+                 defaultNodeId);
     return;
   }
 
   g_autoptr(WpNode) node = static_cast<WpNode*>(
       wp_object_manager_lookup(self->om_, WP_TYPE_NODE, WP_CONSTRAINT_TYPE_G_PROPERTY, "bound-id",
-                               "=u", default_node_id, NULL));
+                               "=u", defaultNodeId, nullptr));
 
-  if (!node) {
+  if (node == nullptr) {
     spdlog::warn("[{}]: (onDefaultNodesApiChanged) - Object with id {} not found", self->name_,
-                 default_node_id);
+                 defaultNodeId);
     return;
   }
 
-  const gchar* default_node_name =
+  const gchar* defaultNodeName =
       wp_pipewire_object_get_property(WP_PIPEWIRE_OBJECT(node), "node.name");
 
   spdlog::debug(
       "[{}]: (onDefaultNodesApiChanged) - got the following default node: Node(name: {}, id: {})",
-      self->name_, default_node_name, default_node_id);
+      self->name_, defaultNodeName, defaultNodeId);
 
-  if (g_strcmp0(self->default_node_name_, default_node_name) == 0) {
+  if (g_strcmp0(self->default_node_name_, defaultNodeName) == 0) {
     spdlog::debug(
         "[{}]: (onDefaultNodesApiChanged) - Default node has not changed. Node(name: {}, id: {}). "
         "Ignoring.",
-        self->name_, self->default_node_name_, default_node_id);
+        self->name_, self->default_node_name_, defaultNodeId);
     return;
   }
 
   spdlog::debug(
       "[{}]: (onDefaultNodesApiChanged) - Default node changed to -> Node(name: {}, id: {})",
-      self->name_, default_node_name, default_node_id);
+      self->name_, defaultNodeName, defaultNodeId);
 
   g_free(self->default_node_name_);
-  self->default_node_name_ = g_strdup(default_node_name);
-  self->node_id_ = default_node_id;
-  updateVolume(self, default_node_id);
-  updateNodeName(self, default_node_id);
+  self->default_node_name_ = g_strdup(defaultNodeName);
+  self->node_id_ = defaultNodeId;
+  updateVolume(self, defaultNodeId);
+  updateNodeName(self, defaultNodeId);
 }
 
 void waybar::modules::Wireplumber::onObjectManagerInstalled(waybar::modules::Wireplumber* self) {
@@ -190,14 +187,14 @@ void waybar::modules::Wireplumber::onObjectManagerInstalled(waybar::modules::Wir
 
   self->def_nodes_api_ = wp_plugin_find(self->wp_core_, "default-nodes-api");
 
-  if (!self->def_nodes_api_) {
+  if (self->def_nodes_api_ == nullptr) {
     spdlog::error("[{}]: default nodes api is not loaded.", self->name_);
     throw std::runtime_error("Default nodes API is not loaded\n");
   }
 
   self->mixer_api_ = wp_plugin_find(self->wp_core_, "mixer-api");
 
-  if (!self->mixer_api_) {
+  if (self->mixer_api_ == nullptr) {
     spdlog::error("[{}]: mixer api is not loaded.", self->name_);
     throw std::runtime_error("Mixer api is not loaded\n");
   }
@@ -206,7 +203,7 @@ void waybar::modules::Wireplumber::onObjectManagerInstalled(waybar::modules::Wir
                         &self->default_node_name_);
   g_signal_emit_by_name(self->def_nodes_api_, "get-default-node", "Audio/Sink", &self->node_id_);
 
-  if (self->default_node_name_) {
+  if (self->default_node_name_ != nullptr) {
     spdlog::debug("[{}]: (onObjectManagerInstalled) - default configured node name: {} and id: {}",
                   self->name_, self->default_node_name_, self->node_id_);
   }
@@ -221,11 +218,11 @@ void waybar::modules::Wireplumber::onObjectManagerInstalled(waybar::modules::Wir
 
 void waybar::modules::Wireplumber::onPluginActivated(WpObject* p, GAsyncResult* res,
                                                      waybar::modules::Wireplumber* self) {
-  auto plugin_name = wp_plugin_get_name(WP_PLUGIN(p));
-  spdlog::debug("[{}]: onPluginActivated: {}", self->name_, plugin_name);
-  g_autoptr(GError) error = NULL;
+  const auto* pluginName = wp_plugin_get_name(WP_PLUGIN(p));
+  spdlog::debug("[{}]: onPluginActivated: {}", self->name_, pluginName);
+  g_autoptr(GError) error = nullptr;
 
-  if (!wp_object_activate_finish(p, res, &error)) {
+  if (wp_object_activate_finish(p, res, &error) == 0) {
     spdlog::error("[{}]: error activating plugin: {}", self->name_, error->message);
     throw std::runtime_error(error->message);
   }
@@ -240,7 +237,7 @@ void waybar::modules::Wireplumber::activatePlugins() {
   for (uint16_t i = 0; i < apis_->len; i++) {
     WpPlugin* plugin = static_cast<WpPlugin*>(g_ptr_array_index(apis_, i));
     pending_plugins_++;
-    wp_object_activate(WP_OBJECT(plugin), WP_PLUGIN_FEATURE_ENABLED, NULL,
+    wp_object_activate(WP_OBJECT(plugin), WP_PLUGIN_FEATURE_ENABLED, nullptr,
                        (GAsyncReadyCallback)onPluginActivated, this);
   }
 }
@@ -248,34 +245,67 @@ void waybar::modules::Wireplumber::activatePlugins() {
 void waybar::modules::Wireplumber::prepare() {
   spdlog::debug("[{}]: preparing object manager", name_);
   wp_object_manager_add_interest(om_, WP_TYPE_NODE, WP_CONSTRAINT_TYPE_PW_PROPERTY, "media.class",
-                                 "=s", "Audio/Sink", NULL);
+                                 "=s", "Audio/Sink", nullptr);
 }
 
-void waybar::modules::Wireplumber::loadRequiredApiModules() {
-  spdlog::debug("[{}]: loading required modules", name_);
-  g_autoptr(GError) error = NULL;
+void waybar::modules::Wireplumber::onDefaultNodesApiLoaded(WpObject* p, GAsyncResult* res,
+                                                           waybar::modules::Wireplumber* self) {
+  gboolean success = FALSE;
+  g_autoptr(GError) error = nullptr;
 
-  if (!wp_core_load_component(wp_core_, "libwireplumber-module-default-nodes-api", "module", NULL,
-                              &error)) {
+  spdlog::debug("[{}]: callback loading default node api module", self->name_);
+
+  success = wp_core_load_component_finish(self->wp_core_, res, &error);
+
+  if (success == FALSE) {
+    spdlog::error("[{}]: default nodes API load failed", self->name_);
+    throw std::runtime_error(error->message);
+  }
+  spdlog::debug("[{}]: loaded default nodes api", self->name_);
+  g_ptr_array_add(self->apis_, wp_plugin_find(self->wp_core_, "default-nodes-api"));
+
+  spdlog::debug("[{}]: loading mixer api module", self->name_);
+  wp_core_load_component(self->wp_core_, "libwireplumber-module-mixer-api", "module", nullptr,
+                         "mixer-api", nullptr, (GAsyncReadyCallback)onMixerApiLoaded, self);
+}
+
+void waybar::modules::Wireplumber::onMixerApiLoaded(WpObject* p, GAsyncResult* res,
+                                                    waybar::modules::Wireplumber* self) {
+  gboolean success = FALSE;
+  g_autoptr(GError) error = nullptr;
+
+  success = wp_core_load_component_finish(self->wp_core_, res, nullptr);
+
+  if (success == FALSE) {
+    spdlog::error("[{}]: mixer API load failed", self->name_);
     throw std::runtime_error(error->message);
   }
 
-  if (!wp_core_load_component(wp_core_, "libwireplumber-module-mixer-api", "module", NULL,
-                              &error)) {
-    throw std::runtime_error(error->message);
-  }
-
-  g_ptr_array_add(apis_, wp_plugin_find(wp_core_, "default-nodes-api"));
-  g_ptr_array_add(apis_, ({
-                    WpPlugin* p = wp_plugin_find(wp_core_, "mixer-api");
-                    g_object_set(G_OBJECT(p), "scale", 1 /* cubic */, NULL);
+  spdlog::debug("[{}]: loaded mixer API", self->name_);
+  g_ptr_array_add(self->apis_, ({
+                    WpPlugin* p = wp_plugin_find(self->wp_core_, "mixer-api");
+                    g_object_set(G_OBJECT(p), "scale", 1 /* cubic */, nullptr);
                     p;
                   }));
+
+  self->activatePlugins();
+
+  self->dp.emit();
+
+  self->event_box_.add_events(Gdk::SCROLL_MASK | Gdk::SMOOTH_SCROLL_MASK);
+  self->event_box_.signal_scroll_event().connect(sigc::mem_fun(*self, &Wireplumber::handleScroll));
+}
+
+void waybar::modules::Wireplumber::asyncLoadRequiredApiModules() {
+  spdlog::debug("[{}]: loading default nodes api module", name_);
+  wp_core_load_component(wp_core_, "libwireplumber-module-default-nodes-api", "module", nullptr,
+                         "default-nodes-api", nullptr, (GAsyncReadyCallback)onDefaultNodesApiLoaded,
+                         this);
 }
 
 auto waybar::modules::Wireplumber::update() -> void {
   auto format = format_;
-  std::string tooltip_format;
+  std::string tooltipFormat;
 
   if (muted_) {
     format = config_["format-muted"].isString() ? config_["format-muted"].asString() : format;
@@ -292,12 +322,12 @@ auto waybar::modules::Wireplumber::update() -> void {
   getState(vol);
 
   if (tooltipEnabled()) {
-    if (tooltip_format.empty() && config_["tooltip-format"].isString()) {
-      tooltip_format = config_["tooltip-format"].asString();
+    if (tooltipFormat.empty() && config_["tooltip-format"].isString()) {
+      tooltipFormat = config_["tooltip-format"].asString();
     }
 
-    if (!tooltip_format.empty()) {
-      label_.set_tooltip_text(fmt::format(fmt::runtime(tooltip_format),
+    if (!tooltipFormat.empty()) {
+      label_.set_tooltip_text(fmt::format(fmt::runtime(tooltipFormat),
                                           fmt::arg("node_name", node_name_),
                                           fmt::arg("volume", vol), fmt::arg("icon", getIcon(vol))));
     } else {
@@ -317,31 +347,31 @@ bool waybar::modules::Wireplumber::handleScroll(GdkEventScroll* e) {
   if (dir == SCROLL_DIR::NONE) {
     return true;
   }
-  double max_volume = 1;
+  double maxVolume = 1;
   double step = 1.0 / 100.0;
   if (config_["scroll-step"].isDouble()) {
     step = config_["scroll-step"].asDouble() / 100.0;
   }
   if (config_["max-volume"].isDouble()) {
-    max_volume = config_["max-volume"].asDouble() / 100.0;
+    maxVolume = config_["max-volume"].asDouble() / 100.0;
   }
 
   if (step < min_step_) step = min_step_;
 
-  double new_vol = volume_;
+  double newVol = volume_;
   if (dir == SCROLL_DIR::UP) {
-    if (volume_ < max_volume) {
-      new_vol = volume_ + step;
-      if (new_vol > max_volume) new_vol = max_volume;
+    if (volume_ < maxVolume) {
+      newVol = volume_ + step;
+      if (newVol > maxVolume) newVol = maxVolume;
     }
   } else if (dir == SCROLL_DIR::DOWN) {
     if (volume_ > 0) {
-      new_vol = volume_ - step;
-      if (new_vol < 0) new_vol = 0;
+      newVol = volume_ - step;
+      if (newVol < 0) newVol = 0;
     }
   }
-  if (new_vol != volume_) {
-    GVariant* variant = g_variant_new_double(new_vol);
+  if (newVol != volume_) {
+    GVariant* variant = g_variant_new_double(newVol);
     gboolean ret;
     g_signal_emit_by_name(mixer_api_, "set-volume", node_id_, variant, &ret);
   }

